File: python/src/anonrouter_confidential/verify/checks.py
"""Shared fail-closed check primitives + verdict assembly.

Kept tiny and pure so every provider verifier composes the SAME semantics: a
required check that fails forces overall status "failed", and the derived
verification level is never stronger than the checks justify.

These are the same check semantics AnonRouter's gateway applies server-side, so an
independent verdict and the gateway's own verdict agree by construction.
"""

from __future__ import annotations

import re
import time
from datetime import datetime, timezone
from typing import Any

from .types import (
    AttestationCheck,
    AttestationExpectations,
    NormalizedVerdict,
    VerificationLevel,
)

DEFAULT_EVIDENCE_MAX_AGE_MS = 5 * 60_000
DEFAULT_CACHE_TTL_MS = 5 * 60_000

_HEX_ONLY = re.compile(r"^[0-9a-f]+$")
_NONCE_SHAPE = re.compile(r"^[0-9a-f]{32,128}$", re.IGNORECASE)


def check(name: str, passed: bool, required: bool, detail: str | None = None) -> AttestationCheck:
    return AttestationCheck(name=name, passed=passed, required=required, detail=detail)


def _now_ms(expectations: AttestationExpectations) -> float:
    return expectations.now_ms if expectations.now_ms is not None else time.time() * 1000.0


class Envelope:
    """Defensively-read attestation envelope."""

    __slots__ = ("endpoint_identity", "fetched_at_ms", "payload")

    def __init__(self, fetched_at_ms: float, endpoint_identity: str, payload: Any) -> None:
        self.fetched_at_ms = fetched_at_ms
        self.endpoint_identity = endpoint_identity
        self.payload = payload


def read_envelope(evidence: Any, expectations: AttestationExpectations) -> Envelope:
    """A bare provider payload (no envelope) is tolerated: fetch time defaults to
    now and endpoint to the expectations."""
    if isinstance(evidence, dict) and "payload" in evidence:
        fetched = evidence.get("fetchedAtMs")
        endpoint = evidence.get("endpointIdentity")
        return Envelope(
            fetched_at_ms=float(fetched) if isinstance(fetched, (int, float)) else _now_ms(expectations),
            endpoint_identity=endpoint if isinstance(endpoint, str) else expectations.endpoint_identity,
            payload=evidence.get("payload"),
        )
    return Envelope(_now_ms(expectations), expectations.endpoint_identity, evidence)


def hex_equal(a: str | None, b: str | None) -> bool:
    """Lowercase-hex equality that does not short-circuit on length. Non-hex,
    empty, or mismatched-length inputs compare unequal."""
    if not a or not b:
        return False
    x = a.lower().removeprefix("0x")
    y = b.lower().removeprefix("0x")
    if len(x) != len(y) or len(x) == 0:
        return False
    if not _HEX_ONLY.match(x) or not _HEX_ONLY.match(y):
        return False
    diff = 0
    for cx, cy in zip(x, y):
        diff |= ord(cx) ^ ord(cy)
    return diff == 0


def all_required_passed(checks: list[AttestationCheck]) -> bool:
    return all((not c.required) or c.passed for c in checks)


def first_required_failure(checks: list[AttestationCheck]) -> str | None:
    for c in checks:
        if c.required and not c.passed:
            return c.name
    return None


def nonce_binding_check(expected_nonce: str, bound_nonce: str | None) -> AttestationCheck:
    if not _NONCE_SHAPE.match(expected_nonce or ""):
        return check("nonce_binding", False, True, "caller nonce missing or too short")
    return check(
        "nonce_binding",
        hex_equal(expected_nonce, bound_nonce),
        True,
        None if bound_nonce else "evidence bound no nonce",
    )


def freshness_check(evidence_epoch_ms: float | None, expectations: AttestationExpectations) -> AttestationCheck:
    now = _now_ms(expectations)
    max_age = expectations.max_evidence_age_ms if expectations.max_evidence_age_ms is not None else DEFAULT_EVIDENCE_MAX_AGE_MS
    if evidence_epoch_ms is None:
        return check("evidence_freshness", False, True, "evidence carried no verifiable timestamp")
    age = now - evidence_epoch_ms
    passed = -30_000 <= age <= max_age
    return check("evidence_freshness", passed, True, None if passed else "evidence outside freshness window")


def resolve_level(requested_level: VerificationLevel, checks: list[AttestationCheck]) -> VerificationLevel:
    if not all_required_passed(checks):
        return "unverified"
    return requested_level


def _iso(ms: float) -> str:
    return datetime.fromtimestamp(ms / 1000.0, tz=timezone.utc).isoformat().replace("+00:00", "Z")


def assemble_result(
    *,
    expectations: AttestationExpectations,
    hardware_type: str,
    requested_level: VerificationLevel,
    measurement_identities: dict[str, str],
    model_weight_identity: str | None,
    attested_tls_spki: str | None,
    attested_encryption_key: str | None,
    attested_signing_key: str | None,
    bound_nonce: str | None,
    verifier_version: str,
    supports_client_opaque_e2ee: bool,
    checks: list[AttestationCheck],
) -> NormalizedVerdict:
    """Assemble the normalized verdict, deriving status/level/expiry from the
    checks. The requested level is CLAMPED: it is never emitted stronger than the
    checks justify, and callers never pass ``hardware-verified``."""
    now = _now_ms(expectations)
    ttl = expectations.cache_ttl_ms if expectations.cache_ttl_ms is not None else DEFAULT_CACHE_TTL_MS
    ok = all_required_passed(checks)
    level = resolve_level(requested_level, checks)
    policy = expectations.measurement_policy or {}
    return NormalizedVerdict(
        status="ok" if ok else "failed",
        verification_level=level,
        privacy_modality=expectations.privacy_modality,
        hardware_type=hardware_type,
        measurement_identities=measurement_identities,
        model_weight_identity=model_weight_identity,
        attested_tls_spki=attested_tls_spki,
        attested_encryption_key=attested_encryption_key,
        attested_signing_key=attested_signing_key,
        nonce=bound_nonce,
        policy_source=policy.get("source") if isinstance(policy, dict) else None,
        verifier_version=verifier_version,
        supports_client_opaque_e2ee=supports_client_opaque_e2ee,
        reason=None if ok else (first_required_failure(checks) or "verification_failed"),
        checks=checks,
        verified_at=_iso(now),
        expires_at=_iso(now + ttl),
    )


def failed_result(
    expectations: AttestationExpectations,
    reason: str,
    verifier_version: str,
    *,
    hardware_type: str = "unknown",
    supports_client_opaque_e2ee: bool = False,
) -> NormalizedVerdict:
    """A fully-failed verdict for the unsupported/error path. Never raises."""
    return assemble_result(
        expectations=expectations,
        hardware_type=hardware_type,
        requested_level="unverified",
        measurement_identities={},
        model_weight_identity=None,
        attested_tls_spki=None,
        attested_encryption_key=None,
        attested_signing_key=None,
        bound_nonce=None,
        verifier_version=verifier_version,
        supports_client_opaque_e2ee=supports_client_opaque_e2ee,
        checks=[check("evidence_present", False, True, reason)],
    )

File: js/confidential/src/verify/checks.ts
// Shared fail-closed check primitives and result assembly for the provider-neutral
// verifier. These are the same check semantics AnonRouter's gateway applies
// server-side, so an independent verdict and the gateway's own verdict agree by
// construction rather than by coincidence. Kept tiny and pure so every
// provider verifier composes the SAME semantics: a required check that fails forces
// overall status "failed", and the derived level is never stronger than justified.

import { hexEqual } from "./crypto.js";
import type {
  AttestationCheck,
  AttestationEnvelope,
  AttestationExpectations,
  HardwareType,
  NormalizedAttestationResult,
  PrivacyModality,
  VerificationLevel
} from "./types.js";

export const DEFAULT_EVIDENCE_MAX_AGE_MS = 5 * 60_000;
export const DEFAULT_CACHE_TTL_MS = 5 * 60_000;

/** Build a check result. `detail` must be safe/content-free. */
export function check(name: string, passed: boolean, required: boolean, detail?: string): AttestationCheck {
  return { name, required, passed, detail };
}

/** Defensively read an AttestationEnvelope. A bare provider payload (no envelope)
 *  is tolerated: fetch time defaults to now and endpoint to the expectations. */
export function readEnvelope(evidence: unknown, expectations: AttestationExpectations): AttestationEnvelope {
  if (evidence && typeof evidence === "object" && "payload" in (evidence as Record<string, unknown>)) {
    const env = evidence as Partial<AttestationEnvelope>;
    return {
      fetchedAtMs: typeof env.fetchedAtMs === "number" ? env.fetchedAtMs : (expectations.now ?? Date.now()),
      endpointIdentity: typeof env.endpointIdentity === "string" ? env.endpointIdentity : expectations.endpointIdentity,
      payload: env.payload
    };
  }
  return { fetchedAtMs: expectations.now ?? Date.now(), endpointIdentity: expectations.endpointIdentity, payload: evidence };
}

/** True when every REQUIRED check passed. Non-required checks are advisory. */
export function allRequiredPassed(checks: AttestationCheck[]): boolean {
  return checks.every((c) => !c.required || c.passed);
}

/** The first failed required check's name (the sanitized failure reason). */
export function firstRequiredFailure(checks: AttestationCheck[]): string | null {
  const failed = checks.find((c) => c.required && !c.passed);
  return failed ? failed.name : null;
}

/**
 * Nonce freshness: the expectations nonce must be non-empty hex of adequate entropy
 * and must equal the nonce the evidence bound. This is the anti-replay anchor; a
 * stale or absent nonce binding is always a required failure.
 */
export function nonceBindingCheck(expectedNonce: string, boundNonce: string | null): AttestationCheck {
  const validShape = /^[0-9a-f]{32,128}$/i.test(expectedNonce);
  if (!validShape) {
    return check("nonce_binding", false, true, "caller nonce missing or too short");
  }
  return check("nonce_binding", hexEqual(expectedNonce, boundNonce), true, boundNonce ? undefined : "evidence bound no nonce");
}

/**
 * Freshness/expiry: the evidence's generation instant must be within the allowed
 * age window relative to `now`. A future-dated or too-old quote fails closed.
 */
export function freshnessCheck(evidenceEpochMs: number | null, expectations: AttestationExpectations): AttestationCheck {
  const now = expectations.now ?? Date.now();
  const maxAge = expectations.maxEvidenceAgeMs ?? DEFAULT_EVIDENCE_MAX_AGE_MS;
  if (evidenceEpochMs === null || !Number.isFinite(evidenceEpochMs)) {
    return check("evidence_freshness", false, true, "evidence carried no verifiable timestamp");
  }
  const age = now - evidenceEpochMs;
  // Allow small clock skew into the future (30s), reject stale beyond maxAge.
  const passed = age >= -30_000 && age <= maxAge;
  return check("evidence_freshness", passed, true, passed ? undefined : "evidence outside freshness window");
}

/** Resolve the overall verification level, clamped by the check outcomes: any
 *  required-check failure caps the level at "unverified". */
export function resolveLevel(requestedLevel: VerificationLevel, checks: AttestationCheck[]): VerificationLevel {
  if (!allRequiredPassed(checks)) return "unverified";
  return requestedLevel;
}

export interface AssembleInput {
  expectations: AttestationExpectations;
  hardwareType: HardwareType;
  requestedLevel: VerificationLevel;
  privacyModality: PrivacyModality;
  measurementIdentities: Record<string, string>;
  modelWeightIdentity: string | null;
  attestedTlsSpki: string | null;
  attestedEncryptionKey: string | null;
  attestedSigningKey: string | null;
  boundNonce: string | null;
  verifierVersion: string;
  supportsClientOpaqueE2ee: boolean;
  checks: AttestationCheck[];
}

/** Assemble the normalized result, deriving status/level/expiry from the checks. */
export function assembleResult(input: AssembleInput): NormalizedAttestationResult {
  const now = input.expectations.now ?? Date.now();
  const ttl = input.expectations.cacheTtlMs ?? DEFAULT_CACHE_TTL_MS;
  const ok = allRequiredPassed(input.checks);
  const level = resolveLevel(input.requestedLevel, input.checks);
  return {
    provider: input.expectations.provider,
    canonicalModel: input.expectations.canonicalModel,
    upstreamModel: input.expectations.upstreamModel,
    routeId: input.expectations.routeId,
    endpointIdentity: input.expectations.endpointIdentity,
    hardwareType: input.hardwareType,
    verificationLevel: level,
    privacyModality: input.privacyModality,
    measurementIdentities: input.measurementIdentities,
    modelWeightIdentity: input.modelWeightIdentity,
    attestedTlsSpki: input.attestedTlsSpki,
    attestedEncryptionKey: input.attestedEncryptionKey,
    attestedSigningKey: input.attestedSigningKey,
    nonce: input.boundNonce,
    verifiedAt: new Date(now).toISOString(),
    expiresAt: new Date(now + ttl).toISOString(),
    policySource: input.expectations.measurementPolicy?.source ?? null,
    verifierVersion: input.verifierVersion,
    supportsClientOpaqueE2ee: input.supportsClientOpaqueE2ee,
    status: ok ? "ok" : "failed",
    reason: ok ? null : firstRequiredFailure(input.checks) ?? "verification_failed",
    checks: input.checks
  };
}

/** A fully-failed result for the unsupported/error path (no evidence, wrong
 *  provider binding, unparseable evidence). Never throws. */
export function failedResult(
  expectations: AttestationExpectations,
  reason: string,
  verifierVersion: string,
  opts: { level?: VerificationLevel; hardwareType?: HardwareType; supportsClientOpaqueE2ee?: boolean } = {}
): NormalizedAttestationResult {
  return assembleResult({
    expectations,
    hardwareType: opts.hardwareType ?? "unknown",
    requestedLevel: opts.level ?? "unverified",
    privacyModality: expectations.privacyModality,
    measurementIdentities: {},
    modelWeightIdentity: null,
    attestedTlsSpki: null,
    attestedEncryptionKey: null,
    attestedSigningKey: null,
    boundNonce: null,
    verifierVersion,
    supportsClientOpaqueE2ee: opts.supportsClientOpaqueE2ee ?? false,
    checks: [check("evidence_present", false, true, reason)]
  });
}

export { hexEqual };
